packages/swingset-liveslots/src/virtualObjectManager.js

```
@@ -15,137 +15,273 @@ const { details: X, quote: q } = assert;
+ *
+ * 'capdatas' is a mutable Object (record) with state property names as keys, and
+ * their capdata { body, slots } as values, and 'valueMap' is a Map with state
+ * property names as keys, and their unmarshalled values as values. The capdata
+ * values can be hardened, but we need the 'capdatas' record to be mutable
+ * because we will modify its contents in place during setters, to retain the
+ * insertion order later (during flush). We need capdata at all so we can
+ * compare the slots before and after the update, to adjust the refcounts. Only
+ * the values of 'valueMap' are exposed to userspace.
+ */
+function makeDataCache(syscall, unserialize) {
+  /** @type {(baseRef: string) => { capdatas: any, valueMap: Map<string, any> }} */
+  const readBacking = baseRef => {
+    const rawState = syscall.vatstoreGet(`vom.${baseRef}`);
+    assert(rawState);
+    const capdatas = JSON.parse(rawState);
+    const valueMap = new Map();
+    for (const prop of Object.getOwnPropertyNames(capdatas)) {
+      valueMap.set(prop, harden(unserialize(capdatas[prop])));
+    }
+    return { capdatas, valueMap };
   };
-  return cache;
+  /** @type {(baseRef: string, value: { capdatas: any, valueMap: Map<string, any> }) => void} */
+  const writeBacking = (baseRef, value) => {
+    const rawState = JSON.stringify(value.capdatas);
+    syscall.vatstoreSet(`vom.${baseRef}`, rawState);
+  };
+  /** @type {(collectionID: string) => void} */
+  const deleteBacking = baseRef => syscall.vatstoreDelete(`vom.${baseRef}`);
+  return makeCache(readBacking, writeBacking, deleteBacking);
+}
+
+function makeContextCache(makeState, makeContext) {
+  // non-writeback cache for "context" objects, { state, self/facets }
+  const readBacking = baseRef => {
+    const state = makeState(baseRef);
+    const context = makeContext(baseRef, state);
+    return context;
+  };
+  const writeBacking = _baseRef => undefined; // never called
+  const deleteBacking = _baseRef => undefined; // never called
```

Should these `Fail` rather than silently no-op?

---

yeah, good idea